daemon/api_notices_test.go

```diff
@@ -0,0 +1,285 @@
+	st.Lock()
+	addNotice(c, st, state.ChangeUpdateNotice, "123", nil)
+	time.Sleep(time.Microsecond)
+	addNotice(c, st, state.WarningNotice, "danger", nil)
+	st.Unlock()
+
+	req, err := http.NewRequest("GET", "/v2/notices?types=change-update&types=warning", nil)
+	c.Assert(err, IsNil)
+	rsp := s.syncReq(c, req, nil)
+	c.Check(rsp.Status, Equals, 200)
+
+	notices, ok := rsp.Result.([]*state.Notice)
+	c.Assert(ok, Equals, true)
+	c.Assert(notices, HasLen, 2)
+	n := noticeToMap(c, notices[0])
+	c.Assert(n["type"], Equals, "change-update")
+	n = noticeToMap(c, notices[1])
+	c.Assert(n["type"], Equals, "warning")
+}
+
+func (s *noticesSuite) TestNoticesFilterMultipleKeys(c *C) {
+	s.daemon(c)
+
+	st := s.d.Overlord().State()
+	st.Lock()
+	addNotice(c, st, state.ChangeUpdateNotice, "123", nil)
+	time.Sleep(time.Microsecond)
+	addNotice(c, st, state.ChangeUpdateNotice, "456", nil)
+	time.Sleep(time.Microsecond)
+	addNotice(c, st, state.WarningNotice, "danger", nil)
+	st.Unlock()
+
+	req, err := http.NewRequest("GET", "/v2/notices?keys=456&keys=danger", nil)
+	c.Assert(err, IsNil)
+	rsp := s.syncReq(c, req, nil)
+	c.Check(rsp.Status, Equals, 200)
+
+	notices, ok := rsp.Result.([]*state.Notice)
+	c.Assert(ok, Equals, true)
+	c.Assert(notices, HasLen, 2)
+	n := noticeToMap(c, notices[0])
+	c.Assert(n["key"], Equals, "456")
+	n = noticeToMap(c, notices[1])
+	c.Assert(n["key"], Equals, "danger")
+}
+
+func (s *noticesSuite) TestNoticesWait(c *C) {
+	s.daemon(c)
+
+	st := s.d.Overlord().State()
+	go func() {
+		time.Sleep(10 * time.Millisecond)
+		st.Lock()
+		addNotice(c, st, state.WarningNotice, "foo", nil)
+		st.Unlock()
+	}()
+
+	req, err := http.NewRequest("GET", "/v2/notices?timeout=1s", nil)
+	c.Assert(err, IsNil)
+	rsp := s.syncReq(c, req, nil)
+	c.Check(rsp.Status, Equals, 200)
+
+	notices, ok := rsp.Result.([]*state.Notice)
+	c.Assert(ok, Equals, true)
+	c.Assert(notices, HasLen, 1)
+	n := noticeToMap(c, notices[0])
+	c.Check(n["type"], Equals, "warning")
+	c.Check(n["key"], Equals, "foo")
+}
+
+func (s *noticesSuite) TestNoticesTimeout(c *C) {
+	s.daemon(c)
+
+	req, err := http.NewRequest("GET", "/v2/notices?timeout=1ms", nil)
+	c.Assert(err, IsNil)
+	rsp := s.syncReq(c, req, nil)
+	c.Check(rsp.Status, Equals, 200)
+
+	notices, ok := rsp.Result.([]*state.Notice)
+	c.Assert(ok, Equals, true)
+	c.Assert(notices, HasLen, 0)
+}
+
+func (s *noticesSuite) TestNoticesRequestCancelled(c *C) {
+	s.daemon(c)
+
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+	go func() {
+		time.Sleep(10 * time.Millisecond)
+		cancel()
+	}()
+
+	start := time.Now()
+	req, err := http.NewRequestWithContext(ctx, "GET", "/v2/notices?timeout=1s", nil)
+	c.Assert(err, IsNil)
+	rsp := s.errorReq(c, req, nil)
+	c.Check(rsp.Status, Equals, 400)
+	c.Check(rsp.Message, Matches, "request canceled")
+
+	elapsed := time.Since(start)
+	c.Check(elapsed > 10*time.Millisecond, Equals, true)
+	c.Check(elapsed < time.Second, Equals, true)
```

Comment:
This feels a bit brittle. It's probably fine but I'm thinking about the slower systems on which spread runs, we've had issues with timing dependent things before I think. Just something to be on the lookout in the first runs

Reply:
I think checking `elapsed > 10*time.Millisecond` is the brittle part, as it could be that the goroutine waiting to cancel the context runs for a while before start time is recorded. I think the solution is to record the start time before spawning the cancel goroutine. Checking that elapsed time is less than 1s should be safe, I would think, on all systems.